File: 940-fruit-into-baskets/fruit-into-baskets.go
func totalFruit(fruits []int) int {
	n := len(fruits)
	lastFruit, secondLastFruit := -1, -1
	lastCount, currMax, max := 0, 0, 0

	for i := 0; i < n; i++ {
		fruit := fruits[i]

		// If the current fruit matches one of the two basket types, increase current max
		if fruit == lastFruit || fruit == secondLastFruit {
			currMax++
		} else {
			// Start new subarray from the last sequence of same fruit
			currMax = lastCount + 1
		}

		// If the current fruit is the same as lastFruit, increase its count
		if fruit == lastFruit {
			lastCount++
		} else {
			// Update basket types
			lastCount = 1
			secondLastFruit = lastFruit
			lastFruit = fruit
		}

		if currMax > max {
			max = currMax
		}
	}

	return max
}

File: 940-fruit-into-baskets/fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        n = len(fruits)
        last_fruit = -1
        second_last_fruit = -1
        last_count = 0
        curr_max = 0
        max_fruits = 0

        for fruit in fruits:
            # Continue collecting if fruit matches one of the two baskets
            if fruit == last_fruit or fruit == second_last_fruit:
                curr_max += 1
            else:
                # Reset current max to last sequence of the same fruit + current one
                curr_max = last_count + 1

            # If same as last fruit, increase its count
            if fruit == last_fruit:
                last_count += 1
            else:
                # Update fruit types and reset count
                last_count = 1
                second_last_fruit = last_fruit
                last_fruit = fruit

            max_fruits = max(max_fruits, curr_max)

        return max_fruits